src/state/dashboard/actions.js

```diff
@@ -0,0 +1,180 @@
+import moment from 'moment';
+
+function getTimePoints(timeRange = 'week', timeUnit = 'day') {
+  const gtDate = moment().startOf(timeRange),
+    timePoints = [];
+  let datePoint = gtDate.clone().startOf(timeUnit);
+
+  while (datePoint.isSameOrBefore(moment())) {
+    timePoints.push(datePoint.clone());
+    datePoint.add(1, timeUnit);
+  }
+
+  return timePoints;
+}
+
+function getTimeUnit(range) {
+  switch (range) {
+    case 'day': {
+      return 'hour';
+    }
+    case 'year': {
+      return 'month';
+    }
+    default: {
+      return 'day';
+    }
+  }
+}
+
+function createTimeQueries(
+  endPoint = '/jobs/survey/',
+  timePoints = [],
+  isCumulative = false,
+  limit = 1000000000
+) {
+  const promiseArray = timePoints.map(async (datePoint, i) => {
+    const gte = isCumulative
+      ? timePoints[0].utc().format()
+      : timePoints[i].utc().format();
+    const lte = timePoints[i + 1]
+      ? timePoints[i + 1].utc().format()
+      : moment()
+          .utc()
+          .format();
+
+    const response = await fetch(
+      `${endPoint}?created_at__gte=${gte}&created_at__lte=${lte}&limit=${limit}`
+    );
+
+    if (response.status === 200) {
+      return await response.json();
+    } else {
+      return [
+        {
+          error: 'there was an error with response'
+        }
+      ];
+    }
+  });
+  return Promise.all(promiseArray);
+}
+
+function getJobStatusesOverTime(jobData = []) {
+  const pending = jobData.map(
```

Comment:
So `pending`, `failed`, `completed` and `open` are all arrays that indicate each job's status? Wouldn't it be better to scan `jobData` once, check each job's `start_time` and/or `success` fields to determine its status? You only need one array as a return value. 

How many elements does `jobData` has typically? If it is not that big, then 1 scan or 4 scans doesn't really matter that much.

Reply:
Suppose `jobData` is a list like this:
```
[ job_group_1, job_group[_2, job_group_3, ... ]
```
You can count the the number of pending/failed/open/complete/ by scanning `jobData` only once, 
instead of scan once to count all jobs that are `pending`, 2nd scan to count `failed` jobs, 3rd scan for `completed`, 4th scan for `open`. (Each call of `map` and `filter` is a scan of the whole list.)

As I mentioned earlier, if the list is short, there is no big deal. But if the list has thousands (or even more) jobs, it may make some difference.

Does this make sense?